gobblin-core/src/main/java/gobblin/writer/partitioner/TimeBasedWriterPartitioner.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.writer.partitioner;
+
+import org.apache.avro.Schema;
+import org.apache.avro.SchemaBuilder;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.commons.lang3.StringUtils;
+import org.joda.time.DateTimeZone;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.State;
+import gobblin.util.ForkOperatorUtils;
+
+
+/**
+ * A {@link WriterPartitioner} that partitions a record based on a timestamp and a {@link DateTimeFormat}.
+ *
+ * A prefix and a suffix can be added to the formatted timestamp, e.g., the partition path can be
+ * 'prefix/2015/11/05/suffix'.
+ *
+ * @author ziliu
+ */
+public abstract class TimeBasedWriterPartitioner<D> implements WriterPartitioner<D> {
```

Comment:
Suggestion:
I think passing in the pattern is pointless, as it will always be of the form yyyy/MM/... until a certain granularity (e.g. no one reasonable would put month first). Instead, users can just set the granularity of their partitioner (i.e. day, hour, minute).
The schema should be created dynamically, so if  a user sets granularity=hour, then the schema would be `{'year':int, 'month':int, 'day':int, 'hour':int}` (possibly plus prefix and suffix). Then the `PartitionedWriter` will construct the path and be able to for example add field names correctly.